src/libcore/intrinsics.rs

```diff
@@ -1478,6 +1506,13 @@ pub unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize) {
 #[stable(feature = "rust1", since = "1.0.0")]
 #[inline]
 pub unsafe fn copy<T>(src: *const T, dst: *mut T, count: usize) {
+    debug_assert!(is_aligned(src), "attempt to copy from unaligned pointer");
+    debug_assert!(is_aligned(dst), "attempt to copy to unaligned pointer");
+    let src_usize = src as usize;
+    let dst_usize = dst as usize;
+    let size = mem::size_of::<T>() * count;
+    debug_assert!(dst_usize < src_usize || dst_usize + size >= src_usize,
+                  "attempt to copy to a destination overlapping with the source");
```

Comment:
Why is this check here? I thought that `copy` allowed overlapping.

Reply:
True, that probably explains the panics.

Also if anything this should call `overlaps`, not re-implement the same logic.

Reply:
So, this checks that `dst` is not in a part of `src` that we haven't read but will. Otherwise, the end of what we read is the beginning of what we write, which is almost certainly an error. If someone actually wants to do that, I will
1) be very surprised and encourage him to rethink
2) suggest to do 2 copy-nonoverlapping instead, so the code explicitly states that the beginning is written twice.

Reply:
Not sure what you mean by "haven't read but will" -- but `copy` is explicitly intended to allow overlapping ranges, so I do not understand why you want to rule them out here.

Reply:
I know what you mean, but for me, it only makes sense when overlapping if `dst` is before `src`. If `dst` is after `src` and they overlap, then when you write the first byte of `src` to the first byte of `dst`, you overwrite a byte of `src` later on (since they overlap). At some point, while reading `src`, you'll find that same byte again, and write it further in `dst`.

For instance, if you have "hello world" in memory, and you copy "hello" onto "lo wo", you'll end up with "helhelherld": 
 - you copy the first 2 bytes -> "helhe world"
 - then you copy the bytes 2-4 -> "helhelherld"

And that's assuming that the bytes are copied left-to-right, which I guess is the case.

Allowing that seems very dangerous, since it's really not obvious that it is what you want, and it doesn't even seem very well specified: if you are copying bytes (no pointer alignment) and the cpu reads the bytes by 8, it could read a byte before it would have been overwritten, leading to a different result than byte-by-byte.

Reply:
This function will do the right thing either way, *and that is its entire point*. In the case you consider dangerous, it'll copy back-to-front.

> For instance, if you have "hello world" in memory, and you copy "hello" onto "lo wo", you'll end up with "helhelherld"

Did you write some code to test this? I think you are wrong, you will end up with "helhellorld", as you would hope.

Reply:
Ah, I was not aware it would copy back-to-front. I'll remove those checks, then.